src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs

```diff
@@ -83,16 +83,19 @@ public static bool IsResumableChangeStreamException(Exception exception)
 
         public static bool IsRetryableWriteException(Exception exception)
         {
+            if (__retryableWriteExceptions.Contains(exception.GetType()))
+            {
+                return true;
+            }
+
             var commandException = exception as MongoCommandException;
             if (commandException != null)
```

Comment:
Because MongoNotPrimaryException is now a MongoCommandException the code was now taking the wrong branch of the if for MongoNotPrimaryException.

The fix is to check the type of the exception first.